3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/PIIRedactor.scala

```diff
@@ -0,0 +1,219 @@
+import org.apache.http.message.BasicNameValuePair
+
+import scala.collection.JavaConverters._
+import scala.util.Try
+import scala.util.control.NonFatal
+
+case class RedactedValue(val replacement: String, val piiType: List[String])
+
+trait PIIRule {
+  def redact(value: String): RedactorResult
+}
+
+/**
+  * Redacts PII matching the full string.
+  * This is typically applied in a context where we suspect what kind of PII may be present.
+  * @param piiType the type of PII matched by the pattern.  For example, NAME or PHONE.
+  * @param valuePattern the pattern describing the PII.  If matched, the value will be redacted.
+  */
+class FullStringPIIDetector(val piiType: String, val valuePattern: Pattern) extends PIIRule {
+
+  def redact(value: String): RedactorResult =
+    if (!valuePattern.matcher(value).matches()) {
+      Left(value)
+    } else {
+      Right(RedactedValue(s"** REDACTED ${piiType.toUpperCase} **", List(piiType)))
+    }
+}
+
+/**
+  * Redacts email addresses.
+  * Since these are fairly recognizable, these can be detected anywhere within a string.
+  * When redacted, a hash of the redacted value will be included with the replacement.
+  * @param valuePattern the pattern to use to recognize email addresses.  Each match will be redacted.
+  */
+class EmailPIIDetector(val valuePattern: Pattern) extends PIIRule {
+
+  val piiType = "EMAIL"
+  val hasher  = Hashing.murmur3_128()
+
+  def redact(value: String): RedactorResult = {
+    val matcher = valuePattern.matcher(value)
+
+    if (matcher.find) {
+      val output = new StringBuffer
+      do {
+        matcher.appendReplacement(output, s"** REDACTED EMAIL [${generateHash(matcher.group)}] **")
+      } while (matcher.find)
+      matcher.appendTail(output)
+      return Right(RedactedValue(output.toString, List(piiType)))
+    }
+    return Left(value)
+  }
+
+  private def generateHash(value: String) =
+    hasher.hashString(value, Charset.forName("UTF-8")).toString
+}
+
+object PIIRedactor {
+
+  type RedactorResult = Either[String, RedactedValue]
+
+  private val anyNonEmptyPattern = Pattern.compile(".+")
+  // Expect at least two digits to avoid false positives for common use case with 0/1 as boolean flag
+  private val phonePattern = Pattern.compile("(?U:[\\d\\s()+\\-]{2,})")
+  private val emailPattern = Pattern.compile(
+    "(?U:\\p{Alnum}[\\p{Alnum}.!#$%&'*+/=?^_`{|}~\\-]+@\\p{Alnum}[\\p{Alnum}\\-]+(\\.[\\p{Alnum}\\-]+)+)")
+
+  private val name: PIIRule     = new FullStringPIIDetector("NAME", anyNonEmptyPattern)
+  private val account: PIIRule  = new FullStringPIIDetector("ACCOUNT", anyNonEmptyPattern)
+  private val password: PIIRule = new FullStringPIIDetector("PASSWORD", anyNonEmptyPattern)
+  private val phone: PIIRule    = new FullStringPIIDetector("PHONE", phonePattern)
+  private val email: PIIRule    = new EmailPIIDetector(emailPattern)
+
+  private val parameterValueRules: Map[String, PIIRule] = Map(
+    "firstname"    -> name,
```

Comment:
do we need to handle camelCase parameter names, or is the uri lower cased before parsing?

Reply:
Yeah, we definitely want case insensitivity.  We lower case the parameter when we do the rule lookup.  Look at lookupParameterRule.  I'm pretty sure there's a test case to cover it too.  Though probably should put a comment here to clarify.